### DISCUSSION TITLE
Create a shorthand function for pipelines

### DISCUSSION THREAD

**Discussion by Shayan-To** [PowerShell/PowerShell#21294] [Q&A] [answered]
I'm using ` | ConvertTo-Csv | Out-File` quite a few times throughout my scripts, and now I want to make a modification to all of them. Is there a way to create a shorthand function for it, so I can refactor my code and do the modifications from one place?

**Comment by MartinGC94** [comment]
Yes, just define a function that takes pipeline input and also does what you want, for example:

```
    function DemoFunction
    {
        Param
        (
            [Parameter(Mandatory, ValueFromPipeline)]
            [System.Object]
            $InputObject
        )
        Process
        {
            $InputObject | ConvertTo-Csv | Out-File
        }
    }
```

Though it's probably worth noting that there already is a wrapper for your use case called `Export-Csv` which handles both the CSV conversion and exporting the data to a file.

**Reply by Shayan-To** [reply]
Unfortunately, this does not work.

I change your code a little bit (to include a random file name) to demonstrate:

```
function DemoFunction
{
    Param
    (
        [Parameter(Mandatory, ValueFromPipeline)]
        [System.Object]
        $InputObject
    )
    Process
    {
        $InputObject | ConvertTo-Csv | Out-File "$(New-Guid).csv"
    }
}
```

Now, if you run `1..10 | % { @{ k1 = 1; k2 = 2 } } | ConvertTo-Csv | Out-File "$(New-Guid).csv"`, it will result in a single file, but if you run `1..10 | % { @{ k1 = 1; k2 = 2 } } | DemoFunction`, it will result in 10 different files, each having a single row (and the header).

**Reply by Shayan-To** [reply]
And thanks about mentioning `Export-Csv`. I need more customizations that it does not provide.

**Reply by MartinGC94** [reply]
You are right, I didn't think that one through. It creates a new pipeline for each item. You could do it like this though:

```
function DemoFunction
{
    Param
    (
        [Parameter(Mandatory, ValueFromPipeline)]
        [System.Object]
        $InputObject
    )
    end
    {
        $input | ConvertTo-Csv | Out-File "$(New-Guid).csv"
    }
}
```

This uses the automatic variable `$input` which contains all the input data that was passed to the function, see: https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_automatic_variables?view=powershell-7.4#input

**Reply by ThomasNieto** [reply]
There is a shorthand to create a function with pipeline support.

```powershell
filter DemoFunction {
  $Input | ConvertTo-Csv | Out-File "$(New-Guid).csv"
}
```

You can also define a parameter too.

```powershell
filter DemoFunction {
  param ($Path)
  $Input | ConvertTo-Csv | Out-File $Path
}
```

**Reply by Shayan-To** [reply]
@ThomasNieto 

Unfortunately, using `$Input` with `Filter` does not do what I want here. If you may, run the examples I provided above, you would see that running `1..10 | % { @{ k1 = 1; k2 = 2 } } | DemoFunction` and running `1..10 | % { @{ k1 = 1; k2 = 2 } } | ConvertTo-Csv | Out-File "$(New-Guid).csv"` yield different results.

The reason is that `Filter`s work like `Function`s with everything inside a `Process` block ([reference](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_functions?view=powershell-7.4#filters)). And behavior of the `$Input` automatic variable is different in `Process` and `End` blocks, as documented in the [link](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_automatic_variables?view=powershell-7.4#input)) @MartinGC94 provided:

> - In the `process` block, the `$input` variable contains the current object in the pipeline.
> - In the `end` block, the `$input` variable enumerates the collection of all input to the function.

So, using `$Input` inside a `Function`, or inside the `End` block of a `Function` would do what I want (as @MartinGC94 taught me), but putting it inside the `Process` block would repeat running the code for each element of the pipeline enumeration, and doesn't have the desired behavior I'm looking for.

BTW, using non of the processing method blocks is just like writing everything inside an `End` block, as is stated [here](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_functions?view=powershell-7.4#input-processing-methods):

> If you don't use a named block, then PowerShell puts the code in the end block of the function.

**Comment by Shayan-To** [comment, marked as answer]
I've been searching for a real solution to this for a very long time (probably years), and even spent some time reading the source code of PowerShell, but didn't find a workable solution. UNTIL I accidentally saw the definition of `mkdir`!

```PowerShell
# mkdir
....

begin {
    $wrappedCmd = $ExecutionContext.InvokeCommand.GetCommand('New-Item', [System.Management.Automation.CommandTypes]::Cmdlet)
    $scriptCmd = {& $wrappedCmd -Type Directory @PSBoundParameters }

    $steppablePipeline = $scriptCmd.GetSteppablePipeline()
    $steppablePipeline.Begin($PSCmdlet)
}

process {
    $steppablePipeline.Process($_)
}

end {
    $steppablePipeline.End()
}
```

This is the real way to solve this problem. Although if it is an option to gather the whole pipline input into an array and then use it, you can use `$input` inside the `End` block or inside the function itself. (This will be slow, and eat up a lot of memory for large pipeline inputs.) (Also see [my previous comment](https://github.com/PowerShell/PowerShell/discussions/21294#discussioncomment-8990033).)

```Powershell
Function DemoFunction1 # Proper way
{
    Begin
    {
        $Pipeline = $Null
        $Pipeline = {
            ConvertTo-Csv | Out-File "$(New-Guid).csv"
        }.GetSteppablePipeline()

        $Pipeline.Begin($True, $ExecutionContext)
    }
    Process
    {
        $Pipeline.Process($_)
    }
    End
    {
        $Pipeline.End()
    }
    Clean
    {
        ${Pipeline}?.Clean()
        ${Pipeline}?.Dispose()
    }
}

Function DemoFunction2 # Slow, memory hungry, but simpler to implement
{
    $input | ConvertTo-Csv | Out-File "$(New-Guid).csv"
}
```
